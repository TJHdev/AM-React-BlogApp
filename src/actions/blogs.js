import database from '../firebase/firebase.js'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// component calls action generator
// action generator returns function
// component dispatches object (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)

export const addBlog = (blog) => ({
  type: 'ADD_BLOG',
  blog: blog
})

export const startAddBlog = (blogData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      body = '',
      createdAt = 0,
      lastEditedAt = 0
    } = blogData;
    
    const blog = { title, body, createdAt, lastEditedAt };

    return database.ref(`users/${uid}/blogs`).push(blog).then((ref) => {
      dispatch(addBlog({
        id: ref.key,
        ...blog
      }));
    });
  };
};

export const removeBlog = ({ id } = {}) => ({
  type: 'REMOVE_BLOG',
  id: id
});

export const startRemoveBlog = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).set(null).then(() => {
      dispatch(removeBlog({ id }));
    });
  }
};

export const editBlog = (id, updates) => ({
  type: 'EDIT_BLOG',
  id: id,
  updates
});

export const startEditBlog = (id, updates = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).update(updates).then(() => {
      dispatch(editBlog(id, updates));
    }); 
  };
};

export const setBlogs = (blogs) => ({
  type: 'SET_BLOGS',
  blogs: blogs
});

export const startSetBlogs = (blogsData = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs`).once('value').then((snapshot) => {
      const blogs = [];
      snapshot.forEach((childSnapshot) => {
        blogs.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

export const startSetBlogsPublic = (blogsData = []) => {
  return (dispatch, getState) => {
    return database.ref('users').once('value').then((snapshot) => {
      const blogs = [];
      const val = snapshot.val();

      Object.keys(val).forEach((childSnapshot) => {
        Object.keys(val[childSnapshot].blogs).forEach((blog) => {
          blogs.push({
            id: blog,
            ...snapshot.val()[childSnapshot].blogs[blog]
          })
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

