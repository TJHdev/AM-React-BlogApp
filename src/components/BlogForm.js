import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();

export default class BlogForm extends React.Component {
  constructor(props) {
    super(props);
    
    this.state = { // no way to do add/edit conditional without constructor
      title: props.blog ? props.blog.title : '',
      body: props.blog ? props.blog.body :  '',
      createdAt: props.blog ? moment(props.blog.createdAt) :  moment(),
      lastEditedAt: props.blog ? moment(props.blog.lastEditedAt) :  moment(),
      calendarFocused: false,
      error: ''
    };
  };
  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };
  onBodyChange = (e) => {
    const body = e.target.value;
    this.setState(() => ({ body }));
    // // alternetive implementation
    // e.persist();
    // this.setState(() => ({ e.target.value }));
  };
  onDateChange = (createdAt) => {
    if (createdAt) { // prevents the user from clearing this value
      this.setState(() => ({ createdAt }));
    }
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.title || !this.state.body) {
      this.setState(() => ({ error: 'Please provide title and body!' }))
    } else {
      this.setState(() => ({ error: '' })) 
      this.props.onSubmit({
        title: this.state.title,
        body: this.state.body,
        createdAt: this.state.createdAt.valueOf(),
        lastEditedAt: moment().valueOf()
      }) // we pass the data up so that the form can remain re-usable
    }
  };
  render() {
    return (
        <form className="form" onSubmit={this.onSubmit}>
          {this.state.error && <p className="form__error">{this.state.error}</p>}
          <input
            type="text"
            placeholder="Title"
            className="text-input"
            autoFocus
            value={this.state.title}
            onChange={this.onTitleChange}
          />
          <textarea
            placeholder="Add the text body"
            className="textarea"
            value={this.state.body}
            onChange={this.onBodyChange}
          ></textarea>
          <div>
            <button className="button">Save Blog</button>
          </div>
        </form>
    );
  }
}