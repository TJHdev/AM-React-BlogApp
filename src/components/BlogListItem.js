import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const BlogListItem = ({ id, title, createdAt }) => (
    <Link className="list-item" to={`/edit/${id}`}>
      <h3 className="list-item__title">{title}</h3>
      <span className="list-item__subtitle">{moment(createdAt).format('MMMM Do, YYYY')}</span>
    </Link>
);

export default BlogListItem;