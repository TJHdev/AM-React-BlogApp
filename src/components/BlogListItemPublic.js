import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const BlogListItemPublic = ({ id, title, createdAt }) => (
  <Link className="list-item" to={`/read/${id}`}>
    <h3 className="list-item__title">{title}</h3>
    <span className="list-item__subtitle">{moment(createdAt).format('h:mma, MMMM Do, YYYY')}</span>
  </Link>
);

export default BlogListItemPublic;