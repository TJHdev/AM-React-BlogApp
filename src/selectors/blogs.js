import moment from 'moment';

// Get visible blogs

const getVisibleBlogs = (blogs, { text, sortBy }) => {
  return blogs.filter((blog) => {
    const textMatch = blog.title.toLowerCase().includes(text.toLowerCase());
    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'dateCreated') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'dateEdited') {
      return a.lastEditedAt < b.lastEditedAt ? 1 : -1;
    } else if (sortBy === 'title') {
      return a.title < b.title ? 1 : -1;
    }
  });
};

export default getVisibleBlogs;