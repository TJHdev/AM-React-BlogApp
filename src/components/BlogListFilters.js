import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { setTextFilter, sortByDateCreated, sortByDateEdited, sortByTitle } from '../actions/filters';

export class BlogListFilters extends React.Component { // this allows us to add state
  state = {
    calendarFocused: null
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'dateCreated') {
      this.props.sortByDateCreated();
    } else if (e.target.value === 'dateEdited') {
      this.props.sortByDateEdited();
    } else if (e.target.value === 'title') {
      this.props.sortByTitle();
    }
  }
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input 
              className="text-input"
              type="text"
              autoFocus
              defaultValue={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder={'Search blogs'} 
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="dateCreated">Date Created</option>
              <option value="dateEdited">Date Last Edited</option>
              <option value="title">Title</option>
            </select>
          </div>
          { this.props.isAuthenticated && 
            <div className="input-group__item">
              <Link className="button" to="/create">Add Blog</Link>
            </div>
          }
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
    isAuthenticated: Boolean(state.auth.uid)
  };
};

const mapDispatchToProps = (dispatch) => (
  {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDateCreated: () => dispatch(sortByDateCreated()),
    sortByDateEdited: () => dispatch(sortByDateEdited()),
    sortByTitle: () => dispatch(sortByTitle()),
  }
)

export default connect(mapStateToProps, mapDispatchToProps)(BlogListFilters);