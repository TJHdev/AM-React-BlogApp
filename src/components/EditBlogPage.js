import React from 'react';
import { connect } from 'react-redux';
import BlogForm from './BlogForm';
import { startEditBlog, startRemoveBlog } from '../actions/blogs';
import RemoveModal from './RemoveModal';

export class EditBlogPage extends React.Component {
  state = {
    confirmRemoveBlog: null
  };
  onSubmit = (blog) => {
    this.props.startEditBlog(this.props.blog.id, blog);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.startRemoveBlog({ id: this.props.blog.id });
    this.props.history.push('/');
  };
  handleOpenRemoveModal = () => {
    this.setState(() => ({ confirmRemoveBlog: true }))
  };
  handleCloseRemoveModal = () => {
    this.setState(() => ({ confirmRemoveBlog: null }))
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit blog</h1>
          </div>
        </div>
        <div className="content-container">
          <BlogForm 
            blog={this.props.blog}
            onSubmit={this.onSubmit}
          />
          <button className="button button--secondary" onClick={this.handleOpenRemoveModal}>Remove Blog</button>
        </div>
        <RemoveModal
          confirmRemoveBlog={this.state.confirmRemoveBlog}
          handleCloseRemoveModal={this.handleCloseRemoveModal}
          onRemove={this.onRemove}
        />
      </div>
    )
  };
}

const mapStateToProps = (state, props) => { // can pass both state and props
  return {
    blog: state.blogs.find((blog) => {
      return blog.id === props.match.params.id;
    })
  };
};

const mapDispatchToProps = (dispatch) => ({
  startEditBlog: (id, blog) => {
    return dispatch(startEditBlog(id, blog));
  },
  startRemoveBlog: (data) => {
    return dispatch(startRemoveBlog(data));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(EditBlogPage);