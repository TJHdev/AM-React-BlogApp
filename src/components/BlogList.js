import React from 'react';
import { connect } from 'react-redux';
import BlogListItem from './BlogListItem';
import selectBlogs from '../selectors/blogs';

export const BlogList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Blogs</div>
      <div className="show-for-desktop">Blog</div>
      <div className="show-for-desktop">Date</div>
    </div>
    <div className="list-body">
      {
        props.blogs.length === 0 ? (
          <div className="list-item list-item--message">
            <span>There are no blogs</span>
          </div>
        ) : (
          props.blogs.map((blog) => (
            <BlogListItem key={blog.id} {...blog} />
          ))
        )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    blogs: selectBlogs(state.blogs, state.filters)
  };
};

export default connect(mapStateToProps)(BlogList);
