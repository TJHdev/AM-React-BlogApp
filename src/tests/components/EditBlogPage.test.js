import React from 'react';
import { shallow } from 'enzyme';
import { EditBlogPage } from '../../components/EditBlogPage';
import blogs from '../fixtures/blogs';

let startEditBlog, handleOpenRemoveModal, history, wrapper;

beforeEach(() => {
  startEditBlog = jest.fn();
  handleOpenRemoveModal = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditBlogPage 
      blog={blogs[1]}
      handleOpenRemoveModal={handleOpenRemoveModal}
      startEditBlog={startEditBlog}
      history={history}
    />
  );
});

test('should render EditBlogPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle startEditBlog', () => {
  wrapper.find('BlogForm').prop('onSubmit')(blogs[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startEditBlog).toHaveBeenLastCalledWith(blogs[1].id, blogs[1]);
});

test('should open the remove modal through change of state', () => {
  expect(wrapper.state('confirmRemoveBlog')).toBe(null);
  wrapper.find('button').prop('onClick')();
  expect(wrapper.state('confirmRemoveBlog')).toBe(true);
});