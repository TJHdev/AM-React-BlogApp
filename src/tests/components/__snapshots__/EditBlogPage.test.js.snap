// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditBlogPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit blog
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <BlogForm
      blog={
        Object {
          "body": "Body bOdy boDy bodY",
          "createdAt": -345600000,
          "id": "2",
          "lastEditedAt": 345600000,
          "title": "are you sure you want to do this?",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Blog
    </button>
  </div>
  <RemoveModal
    confirmRemoveBlog={null}
    handleCloseRemoveModal={[Function]}
    onRemove={[Function]}
  />
</div>
`;
